agent/drive_handler.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package main
+
+import (
+	"io/ioutil"
+	"path/filepath"
+	"strconv"
+	"strings"
+)
+
+const (
+	blockPath = "/sys/block"
+
+	blockMajorMinor = "dev"
+	blockSize       = "size"
+)
+
+type drive struct {
+	Name       string
+	DriveID    string
+	MajorMinor string
+	Size       int
+	Type       string
+}
+
+type driveHandler struct {
+	drives map[string]drive
+}
+
+// DiscoverDrives will use lsblk to populate the drives field in
+// the drive handler.
+func (d *driveHandler) DiscoverDrives() error {
+	names, err := getListOfBlockDeviceNames()
+	if err != nil {
+		return err
+	}
+
+	drives := map[string]drive{}
+	for _, name := range names {
+		d, err := buildDrive(name)
+		if err != nil {
+			return err
+		}
+
+		drives[name] = d
+	}
+
+	for k, v := range drives {
+		if !isStubDrive(v) {
+			continue
+		}
+
+		id, err := getDriveID(v)
+		if err != nil {
+			return err
+		}
+
+		v.DriveID = id
+		drives[k] = v
+	}
+
+	d.drives = drives
+	return nil
+}
+
+func (d driveHandler) GetDrive(id string) (drive, bool) {
+	v, ok := d.drives[id]
+	return v, ok
+}
+
+func getDriveID(d drive) (string, error) {
+	path := filepath.Join("/dev", d.Name)
+	b, err := ioutil.ReadFile(path)
+	return strings.TrimSpace(string(b)), err
+}
+
+// isStubDrive will check to see if a given drive is a stub drive. This is done
+// by looking to see the size of the drive is exactly 512 bytes
```

Comment:
Why is the definition of a stub drive one that is exactly 512 bytes?  What if we want to attach a non-stub drive whose size is 512 bytes?

Reply:
If the stub drive is 1048, only 1024 of it will be visible in the VM since 1048- l1048%512=1024

Reply:
That's not my question though.  You're checking that the size is exactly 512, and I don't think that's necessarily a reasonable indication that this is a "stub" drive.  Instead, I think we should be looking for specific content, like Pascal string- or netstring-formatted data indicating that this is specifically a firecracker-containerd generated stub.

Reply:
@samuelkarp - My concern there is if there is a large file there than reading it could be costly. I don't know if we should be looking at the content because of that. I would be okay with checking the contents of a file within a given range, like anything less than 100kb?

Reply:
That's why I am suggesting a length-encoded format like Pascal strings or netstrings.